Run CopyKAT on Ewings samples

allyhawkins opened the pull request:
### Purpose/implementation Section

#### Please link to the GitHub issue that this pull request addresses.

Addresses #403


#### What is the goal of this pull request?

This PR evaluates using [`CopyKAT`](https://github.com/navinlabcode/copykat) to annotate cells in a Ewing sarcoma sample as tumor or normal cells based on copy number inference. The goal was to see how accurate `CopyKAT` is in assigning cells as aneuploid or tumor cells. I used the annotations obtained from the marker gene analysis (see #391) as the "ground truth" for which cells we expect to be tumor cells. 



#### Briefly describe the general approach you took to achieve this goal.

- `CopyKAT` takes ~1-2 hours when using 4 CPUs with this particular sample, so I included a script to actually run `CopyKAT` outside of the notebook. This script runs `CopyKAT` twice - once with the default parameters and no normal cells and once with the addition of a list of normal cells to use as the reference. 
- For the normal cells, I used cells that were annotated as endothelial cells in both SingleR and CellAssign. This group of cells has almost complete overlap using the two different methods and was one of the normal cell types that lined up with the normal cells we identified in the marker gene analysis. 
- The results from this script are saved to `results/copykat` and then read in as input to the notebook included in this PR that evaluates these results. 
- In the notebook, I compared the results from `CopyKAT` to our manual annotation and also to the cell types assigned by `SingleR` and `CellAssign`. 

#### If known, do you anticipate filing additional pull requests to complete this analysis module?

I plan on looking at additional copy number inference methods in other PRs. 

### Results

#### What is the name of your results bucket on S3?

`s3://researcher-211125375652-us-east-2/ewings-cell-type/results/copykat`

#### What types of results does your code produce (e.g., table, figure)?

The output files from running `CopyKAT` both with and without a normal cell type reference. 

#### What is your summary of the results?

These results are used as input to `03-copykat.Rmd`. 
Generally, `CopyKAT` appears to label the same cells as aneuploid that we label as tumor cells using marker gene analysis. This is shown in a UMAP and also by calculating the confusion matrix between the two annotations. 

We also see that the same cell types from `SingleR` and `CellAssign` that line up with tumor cells in the marker gene analysis line up with aneuploid cells. 

I compared using a reference with `CopyKAT` to not using a reference, and the results are nearly identical with no big differences. 
The biggest difference is in the run time between the two. Running `CopyKAT` with 4 CPUs and no reference took ~ 1.5 - 2 hours and running with the reference only took 45 minutes - 1 hour. 


### Provide directions for reviewers

#### What are the software and computational requirements needed to be able to run the code in this PR?

I recommend running the `run-copykat.R` with at least 4 CPUs to decrease total run time. I haven't officially benchmarked the CPU/ run time relationship but it seemed to save 20-30 minutes of total time when I increased the CPUs. 
There are no special requirements for running the notebook. 


### Author checklists

_Check all those that apply._
_Note that you may find it easier to check off these items after the pull request is actually filed._


#### Analysis module and review

- [x] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [x] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [x] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [ ] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [ ] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [x] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

jashapiro reviewed (commented):
This looks good overall. I have some suggestions about making things a bit more flexible for future explorations (in particular allowing an arbitrary list of normal cells).

There are also a few cleanup things in the notebook and script.

The other thing I would like to see is the start of a bash script (or workflow) for running the workflow, now that there are few steps. The readme is good, but having a record of the exact commands that were run is better! That doesn't mean you have to actually run from the script, but you _could_ and it will make it easier to generate output for future samples, and easier to translate into a workflow later.

Finally, I was wondering if it would be useful to see any more of the output from copyKAT in the notebook. In particular, I was wondering if seeing any of the lower-level CNV output would be useful: is it picking up expected aneuploidies for Ewings? A heatmap of copy number by genomic location for each cell might be good for visualizing that, and I think you can get that out of copyKAT?

allyhawkins commented:
Thanks for the comments and feedback @jashapiro, I've made quite a few changes based on your comments. 

For reference, here's a rendered copy of the notebook: 
[03-copykat.html.zip](https://github.com/AlexsLemonade/OpenScPCA-analysis/files/15278430/03-copykat.html.zip)

> As I said above, I think I would separate this part out to a separate script, or a block in the notebook. The downside, I guess, is you might have to read in the SCE more than once for the different scripts. But I think the flexibility of might be worth it.

I ended up creating a new script `identify-normal-cells.R` that creates and saves a list of barcodes corresponding to "normal cells" for a given library. Right now, it just pulls normal cells from SingleR and/or CellAssign as those are the annotations found in the SCE object, but if we end up adding normal cells from manual annotations to the SCE or want to do this with manual annotations, we can make modifications later. This list is also saved to the repo in `references/normal_cell_lists`, let me know if you think I should move it. 

> For future workflow reasons, if we are doing this more than once, it might make sense to just call the script twice: once with and once without normal cells. For now I think the two versions is fine, but I can imagine you might want to test with more iterations of "normal".

I updated the script to only run copyKAT once and it uses normal cells if provided, but otherwise doesn't use them. I also added a new argument to specify the folder within the copykat results folder to save things. This means results are now in `results/copykat/SCPCLXXX/no_normal/SCPCL*.txt`. 

> The other thing I would like to see is the start of a bash script (or workflow) for running the workflow, now that there are few steps. The readme is good, but having a record of the exact commands that were run is better! That doesn't mean you have to actually run from the script, but you could and it will make it easier to generate output for future samples, and easier to translate into a workflow later.

I definitely agree that we could start turning some of this into a "workflow" of sorts! This is actually something I wanted to do to be able to easily run additional samples, so I'll file an issue for this. I think I want to wrap up the exploratory analysis for this and CellAssign before doing that. 

> Finally, I was wondering if it would be useful to see any more of the output from copyKAT in the notebook. In particular, I was wondering if seeing any of the lower-level CNV output would be useful: is it picking up expected aneuploidies for Ewings? A heatmap of copy number by genomic location for each cell might be good for visualizing that, and I think you can get that out of copyKAT?

To be honest, the output of `CopyKAT` is not the most intuitive and not very well documented. It does output a heatmap of the cnvs, so I just added those to the rendered report for easy viewing together with the other analysis. But they don't label by chromosome... so it's hard to look for specific alterations there. I just can see there's an increase in CNVs in the aneuploid population. 

Regarding looking for CNVs commonly found in Ewing's this is a great idea, but I'm not quite sure how to do that well with the output here. I ended up adding a section where I tried to look at common gains of Chr8, Chr12, and Chr1 and loss of Chr16. To do this I read in the full predictions that include a CNV estimation number for each cell and each gene. I then took the mean value across each chromosome for each cell and plotted that on a UMAP and density. I do see some gain in Chr8 and maybe Chr12, but that's it (although I think Chr8 is the most prevalent?). 

This all has a caveat of I'm not totally sure how these CNV estimation numbers should be interpreted. My understanding is that these are the values used to create the heatmap so above 0 means an increase in copy number for that gene and below 0 means a decrease.

Another thing I want to note is that I dropped the ball and forgot to install pre-commit in my fork, until this week... So I'm now unable to commit the html file without hooks failing. I think I want to just ignore the rendered html anyways and I can save them in results and on S3? What do you think?

allyhawkins commented:
It looks like I included the outdated report, here's the actual current report: 
[03-copykat.html.zip](https://github.com/AlexsLemonade/OpenScPCA-analysis/files/15279015/03-copykat.html.zip)

jashapiro commented:
> Another thing I want to note is that I dropped the ball and forgot to install pre-commit in my fork, until this week... So I'm now unable to commit the html file without hooks failing. I think I want to just ignore the rendered html anyways and I can save them in results and on S3? What do you think?

How big are your html files? I think we do want people to commit those, in general, so if we can set a reasonable limit for the expected size, we probably want to just bump up the limit in pre-commit.

allyhawkins commented on analyses/ewings-cell-type/scripts/run-copykat.R:
```diff
@@ -0,0 +1,99 @@
+  output.seg = FALSE,
+  n.cores = opt$threads
+)
+
```
I don't think there's a way to _not_ save the tables. Also the only thing that gets stored in this object is the contents of the predictions table, so I didn't save it because the predictions file has all that information and is getting saved anyways.

allyhawkins commented on analyses/ewings-cell-type/03-copykat.Rmd:
```diff
@@ -0,0 +1,433 @@
+# read in tumor normal classifications 
+manual_classifications_df <- readr::read_tsv(classifications_file)
+
+# read in ck predictions from both reference types (no_normal and with_normal)
+ck_results_df <- predictions_paths |> 
+  purrr::map(readr::read_tsv) |> 
+  dplyr::bind_rows(.id = "reference_used")
+
+# read in full gene by cell copy number detection results 
+full_ck_results_df <- full_ck_result_paths |> 
+  purrr::map(readr::read_tsv) |> 
+  dplyr::bind_rows(.id = "reference_used")
```
Just reiterating my other reply to your comment that the final RDS object doesn't have the full results... only the final predictions. So we would still have to read this in.

jashapiro replied:
If the object only has the final results, then I guess that is that. I don't think these enormous files are worth saving though, especially if the predictions file ends up being enough, and the rds file is more svelte. 

As for how: instead of moving to the output directory, I would probably move to a temp directory (made with `tempdir()` ) to do the calculations and then copy the output files I actually wanted to the specified output directory. (And then remove the temporary directory to clean up). But I don't think you need to do that now.

allyhawkins commented on analyses/ewings-cell-type/03-copykat.Rmd:
```diff
@@ -0,0 +1,433 @@
+
+### Heatmap without reference
+
+![](`r file.path(results_dir, "no_normal", "SCPCL000822_copykat_heatmap.jpeg")`)
```
Yea I totally agree! I did try to create the plot on my own but it's so large it crashes. I think it would be more beneficial to be able to label isolate specific aneuploides and then plot those, but definitely not worth the effort at this stage. You'll see that the information from InferCNV I think actually does a better job of giving us individual CNV annotation information (and more informative plots).

allyhawkins commented on analyses/ewings-cell-type/03-copykat.Rmd:
````diff
@@ -0,0 +1,433 @@
+
+```{r}
+# for every cell, calculate the mean detection level across all genes in a given chromosome
+full_cnv_df <- full_ck_results_df |> 
````
The prediction table is just two columns - barcode and classification as aneuploid/diploid. There is another large table that is of similar size that does break down the chr into equal segments. I had originally read this in over that one to look for CDKN2A amplifications, but that gene wasn't present so I didn't do anything with that.

jashapiro replied:
I guess what I was thinking is that one, the `$CNAmat` in the output object, according to at least some of the docs?

But if it doesn't have the bands, that would suck.

allyhawkins replied:
From my memory it doesn't have the bands, but when I address #433 I can double check that!

allyhawkins commented on analyses/ewings-cell-type/03-copykat.Rmd:
```diff
@@ -0,0 +1,433 @@
+    end_position, 
+    ensembl_gene_id, 
+    hgnc_symbol, 
+    band),
```
> But I also suspect there is some kind of segmental output you could use in the full data...

One would assume this yes, but there's either this table or the table broken down by equal chromosome segments. That file doesn't give anything beyond chr, start, and end for each region. This is the only file that has band information. I feel like I tried to incorporate band, but it wasn't grouping as I would expect. I'll revisit this.

allyhawkins commented on analyses/ewings-cell-type/03-copykat.Rmd:
```diff
@@ -0,0 +1,433 @@
+    names_to = "barcodes",
+    values_to = "cnv_detection") |> 
+  dplyr::group_by(chromosome_name, barcodes, reference_used) |> 
+  dplyr::summarise(mean_cnv_detection = mean(cnv_detection)) 
```
> Is there any stat for the significance of the cnv_detection? I feel like what we would want here is not the average over all genes but something that takes into account that some are going to be more informative than others.

I looked for stats on the calls and have found nothing... there's really not a lot in the output. This is why I'm hesitant to really interpret these plots too much since all I have is this arbitrary number for detection. Again, there is no "predictions" table, just labeling of each cell as aneuploid and diploid.

allyhawkins commented:
> Okay, this is looking pretty good! I made a couple of suggestions about looking at the copykat results, one of which is that I probably would not use the raw data for plotting/chromosome comparisons, and would instead use the region-level predictions.

I updated this to use the other matrix file, which just has the chromosome, position, and then a column for each barcode with the CNA estimation. I also removed the UMAPs and instead look at CNV estimation across every chromosome. It looks like Chr8 is the most striking gain which is reassuring! 

Here's the updated report with that new plot: 
[03-copykat.html.zip](https://github.com/AlexsLemonade/OpenScPCA-analysis/files/15299497/03-copykat.html.zip)


> Other than that (and removing the umaps for that section), most of what I suggested could probably be put off to refinement PRs if you decided to keep the copykat results part of your predictions. Even if you do not, I think some of this may be useful for validation of tumor cell classifications, but for that I might only use the no-normal data, as it is more independent and worked pretty well in this case at least.

I totally agree about using the no-normal data moving forward! I addressed most other things you mentioned by replying to your comments.

- As for saving the full results as rds and moving the other files to a temp directory, I think this is a good idea, but requires some reorganizing and re-running CopyKAT. In favor of getting this wrapped up and not making this PR bigger, I'm going to file an issue to address that separately. I do think we want to have the ability to run CopyKAT moving forward so this is worth doing to me. 
- For the changes to the select normal cells script, I think everything you mentioned are good options. I do think that's something to do in the future if we feel we need that flexibility. 
- I looked into grouping by bands and not just chromosomes and remember why I didn't keep that in the first place. It's not just p/q bands, it's all 44 segments within the bands. I could pull out the p/q information, but at this point I don't know that we _need_ that. But that information is there.

jashapiro commented on analyses/cell-type-ewings/environment.yml:
This is the one other place I might change: match the name of the environment to the directory.

allyhawkins replied:
Done

jashapiro commented on analyses/cell-type-ewings/03-copykat.Rmd:
````diff
@@ -0,0 +1,402 @@
+
+```{r}
+# for every cell, calculate the mean detection level across all genes in a given chromosome
+full_cnv_df <- full_ck_results_df |> 
+  tidyr::pivot_longer(cols = -c(
````
You might want to remove the `NA` chromosome values somewhere in here. Not sure where they are coming from but they do add a row to your plot. (Maybe Y chromosome genes? I assume it is labelling X as 23)

jashapiro reviewed (approved):
I think this should be fine to go in as is. We certainly may want to refine some things as we build a full report, but I agree that it is better to move along!
